From e4d3d48e2911dbda4149c77b550a5886ef5506cf Mon Sep 17 00:00:00 2001
From: Palmer Dabbelt <palmer@dabbelt.com>
Date: Fri, 2 Nov 2012 09:48:26 -0700
Subject: [PATCH] Add support for short nicknames

Some users (particularly people on Jabber who use their full names)
have very long nicknames.  While it's nice to show these in the title
bar, it's unnecessary to show the full name on every line of the
message.

This adds a configuration option to show only the first N characters
of each buddy's nickname.
---
 pidgin/gtkconv.c  | 21 +++++++++++++++++++++
 pidgin/gtkprefs.c | 21 +++++++++++++++++++++
 2 files changed, 42 insertions(+)

diff --git a/pidgin/gtkconv.c b/pidgin/gtkconv.c
index 4f0121d..621dd16 100644
--- a/pidgin/gtkconv.c
+++ b/pidgin/gtkconv.c
@@ -5761,6 +5761,8 @@ str_embed_direction_chars(char **str)
 #endif
 }
 
+#define LIMIT_NICK_LENGTH_MAX 1024
+
 static void
 pidgin_conv_write_conv(PurpleConversation *conv, const char *name, const char *alias,
 						const char *message, PurpleMessageFlags flags,
@@ -5786,11 +5788,25 @@ pidgin_conv_write_conv(PurpleConversation *conv, const char *name, const char *a
 	char *bracket;
 	int tag_count = 0;
 	gboolean is_rtl_message = FALSE;
+	char short_alias[LIMIT_NICK_LENGTH_MAX];
+	int limit_nick;
+	int limit_nick_length;
 
 	g_return_if_fail(conv != NULL);
 	gtkconv = PIDGIN_CONVERSATION(conv);
 	g_return_if_fail(gtkconv != NULL);
 
+	limit_nick = purple_prefs_get_bool(PIDGIN_PREFS_ROOT "/conversations/limit_nick");
+
+	if (limit_nick && strlen(alias) < LIMIT_NICK_LENGTH_MAX)
+	{
+		limit_nick_length = purple_prefs_get_int(PIDGIN_PREFS_ROOT "/conversations/limit_nick_length");
+		limit_nick_length = (limit_nick_length > LIMIT_NICK_LENGTH_MAX) ? LIMIT_NICK_LENGTH_MAX : limit_nick_length;
+
+		strncpy(short_alias, alias, limit_nick_length);
+		alias = short_alias;
+	}
+
 	if (gtkconv->attach.timer) {
 		/* We are currently in the process of filling up the buffer with the message
 		 * history of the conversation. So we do not need to add the message here.
@@ -7892,6 +7911,8 @@ pidgin_conversations_init(void)
 	purple_prefs_add_bool(PIDGIN_PREFS_ROOT "/conversations/resize_custom_smileys", TRUE);
 	purple_prefs_add_int(PIDGIN_PREFS_ROOT "/conversations/custom_smileys_size", 96);
 	purple_prefs_add_int(PIDGIN_PREFS_ROOT "/conversations/minimum_entry_lines", 2);
+	purple_prefs_add_bool(PIDGIN_PREFS_ROOT "/conversations/limit_nick", FALSE);
+	purple_prefs_add_int(PIDGIN_PREFS_ROOT "/conversations/limit_nick_length", 8);
 
 	purple_prefs_add_bool(PIDGIN_PREFS_ROOT "/conversations/show_timestamps", TRUE);
 	purple_prefs_add_bool(PIDGIN_PREFS_ROOT "/conversations/show_formatting_toolbar", TRUE);
diff --git a/pidgin/gtkprefs.c b/pidgin/gtkprefs.c
index 54381f6..52428d9 100644
--- a/pidgin/gtkprefs.c
+++ b/pidgin/gtkprefs.c
@@ -1438,6 +1438,8 @@ pidgin_custom_font_set(GtkFontButton *font_button, gpointer nul)
 }
 #endif
 
+#define LIMIT_NICK_LENGTH_MAX 1024
+
 static GtkWidget *
 conv_page(void)
 {
@@ -1507,6 +1509,25 @@ conv_page(void)
 		PIDGIN_PREFS_ROOT "/conversations/minimum_entry_lines",
 		1, 8, NULL);
 
+	hbox = gtk_hbox_new(FALSE, PIDGIN_HIG_BOX_SPACE);
+
+	checkbox = pidgin_prefs_checkbox(_("Limit nickname length"),
+			PIDGIN_PREFS_ROOT "/conversations/limit_nick", hbox);
+
+	spin_button = pidgin_prefs_labeled_spin_button(hbox,
+		_("Maximum length:"),
+		PIDGIN_PREFS_ROOT "/conversations/limit_nick_length",
+		1, LIMIT_NICK_LENGTH_MAX, NULL);
+
+	if (!purple_prefs_get_bool(
+				PIDGIN_PREFS_ROOT "/conversations/limit_nick"))
+		gtk_widget_set_sensitive(GTK_WIDGET(spin_button), FALSE);
+
+	g_signal_connect(G_OBJECT(checkbox), "clicked",
+					 G_CALLBACK(pidgin_toggle_sensitive), spin_button);
+
+	pidgin_add_widget_to_vbox(GTK_BOX(vbox), NULL, NULL, hbox, TRUE, NULL);
+
 #ifdef _WIN32
 	{
 	GtkWidget *fontpref, *font_button, *hbox;
-- 
1.7.12.4

